src/interpreter/Exceptions.kt

```diff
@@ -0,0 +1,41 @@
+package interpreter
+
+import core.*
+
+sealed class InterpreterException(msg : String = "", cause : Exception? = null)
+    : RuntimeException(msg, cause)
+
+class TopLevelException(msg : String = "", cause : Exception? = null, val top : AST)
+    : InterpreterException(msg, cause) {
+    override fun toString() = super.toString() + "\nOccurring in $top"
+}
+
+class InvalidPassException(msg : String = "", cause : Exception? = null, val wrongAST : AST, val pass : String)
+    : InterpreterException(msg, cause) {
+    override fun toString() = super.toString() + "\n$wrongAST didn't undergo $pass"
+}
+
+class UndefinedVariableException(msg : String = "", cause : Exception? = null, val wrongAST : AST, val wrongId : String)
+    : InterpreterException(msg, cause) {
+    override fun toString() = super.toString() + "\n$wrongId not defined in $wrongAST"
+}
+
+class IncorrectTypeException(msg : String = "", cause : Exception? = null, val wrongVal : Value)
```

Comment:
would it make sense to have an expectedType argument?

Reply:
Yeah go ahead

Reply:
Once the typecheck is working I'm going to replace all these with `!!` though

Reply:
oh ok makes sense